### DISCUSSION TITLE
ZXing-C++ WebAssembly as an ES module with types

### DISCUSSION THREAD

**Discussion by Sec-ant** [zxing-cpp/zxing-cpp#643] [Show and tell]
Hey @axxel,

Awesome work you're doing! I'm maintaining a repo called [`zxing-wasm`](https://github.com/Sec-ant/zxing-wasm). In it, I've added your repo as a git submodule and rewrote the [`.cpp` code](https://github.com/Sec-ant/zxing-wasm/blob/main/src/cpp/ZXingWasm.cpp) regarding Emscripten bindings. This exposes more options and details like `DecodeHints` and `Result`, making it easier for WebAssembly to interoperate with JavaScript. The library is built on WebAssembly, written in TypeScript, and published as [an ES module on npmjs](https://www.npmjs.com/package/zxing-wasm)—making it easy to integrate into any frontend project.

It's already being used in `barcode-detector` and `vue-qrcode-reader` with good results. Would you be open to adding [`zxing-wasm`](https://github.com/Sec-ant/zxing-wasm) to the Wrappers/Bindings section of your README to give it more visibility?

Thanks a bunch!

**Comment by axxel** [comment]
Hi @Sec-ant,

Thanks for the feedback and thanks for bringing your work to my attention. I had glance over your project and noticed that it is quite a bit code in total and wonder what it offers over the 'plain' WASM wrapper we already have. Here is what I can see so far:
 * no `_malloc` and `HEAPU8`
 * a full list of options via DecodeHints and result properties
 * npm packaging

Can you point to something else? I'm wandering which if any of those advantages could be easily integrated/added to the wrapper we already have.

I obviously already added a link to an external project (flutter wrapper) but there is was something we do not provide directly already.

**Reply by Sec-ant** [reply]
Thanks for your prompt reply, @axxel. Much appreciated.

I think for the demo purpose the 'plain' wrapper is adequate enough and already illustrated the basic usage when dealing with WASM. What I do in [`zxing-wasm`](https://github.com/Sec-ant/zxing-wasm) is to try to make it less painful for web devs to use it. WebAssembly may still be a foreigner as in view from the mental models of many web devs including myself. I just want to hide it from the surface and provide a ready-to-use package while without compromising many of the utilities it can offer. I also want to make it more competitive among many other frontend barcode decoding packages like ZBar.js or some pure js implementations, even commercial sdks.

> Can you point to something else?

Apart from the points you listed, some other tweaks I made include:

- change `--bind` to `-lembind` as the former is said to be [deprecated](https://emscripten.org/docs/tools_reference/emcc.html#emcc-bind)
- set the [`MODULARIZE` and `EXPORT_ES6`](https://github.com/Sec-ant/zxing-wasm/blob/126bacd68a123a3c1eac5f0ec32ff9f86f3487a7/src/cpp/CMakeLists.txt#L35-L37C20) flags. es modules are gradually supported by most of the browsers and runtimes, and are more and more preferred nowadays
- rewrap the "writeBarcode" part to make its API be more align with the "readBarcodes" side. I saw your discussions in other issues and I'm aware that you may redesign the writing part. This is just a temporary attempt because most of the use cases are still focusing on *reading* barcodes
- use TypeScript and publish type declarations as well to provide type checks and type hints for better DX
- also did some conversions for the enum bindings on the TS side, as string unions are generally more [the preferred choices](https://stackoverflow.com/a/60041791), though it can introduce some performance penalties. This is rather opinionated I think

Some of the above tweaks may seem too much and may not be appropriate to be directly added to the existing wrapper. I also find maintaining another repo and using submodules to manage it more acceptable.

> something we do not provide directly already.

Totally understand, I just saw some other porting works are being conducted, so I was wondering if my work could be recognized too and I also want to avoid duplication of effort for myself or others because of lack of communication or visibility. Anyway, I'd love to invest time in this and feel free to contact me if you have other thoughts.

**Comment by axxel** [comment]
I added a short introduction and link to the [WASM/README](https://github.com/zxing-cpp/zxing-cpp/tree/master/wrappers/wasm).

**Reply by Sec-ant** [reply]
Thank you. It means a lot!

**Comment by Sec-ant** [comment]
Just thought maybe you should know, I've also published a demo website to show its function: https://zxing-wasm-demo.netlify.app
